devnotes/pnor_layout.md

```diff
@@ -0,0 +1,161 @@
+```cpp
+ struct ffs_hdr {
+    uint32_t         magic;         // @magic:          Eye catcher/corruption detector
+                                    //                  should always contain 0x50415254 or text PART
+    uint32_t         version;       // @version:        Version of the structure
+    uint32_t         size;          // @size:           Size of partition table (in block_size)
+    uint32_t         entry_size;    // @entry_size:     Size of struct ffs_entry element (in bytes)
+    uint32_t         entry_count;   // @entry_count:    Number of struct ffs_entry elements in @entries array
+    uint32_t         block_size;    // @block_size:     Size of block on device (in bytes)
+    uint32_t         block_count;   // @block_count:    Number of blocks on device
+    uint32_t         resvd[4];      // @resvd:          Reserved words for future use
+    uint32_t         checksum;      // @checksum:       Header checksum
+    struct ffs_entry entries[];     // @entries:        Pointer to array of partition entries
+} __attribute__ ((packed));
+
+struct ffs_entry {
+    char     name[16];      // @name:       Opaque null terminated string
+    uint32_t base;          // @base:       Starting offset of partition in flash (in hdr.block_size)
+    uint32_t size;          // @size:       Partition size (in hdr.block_size)
+    uint32_t pid;           // @pid:        Parent partition entry (0xFFFFFFFF for toplevel)
+    uint32_t id;            // @id:         Partition entry ID [1..65536]
+    uint32_t type;          // @type:       Describe type of partition
+    uint32_t flags;         // @flags:      Partition attributes (optional)
+    uint32_t actual;        // @actual:     Actual partition size (in bytes)
+    uint32_t resvd[4];      // @resvd:      Reserved words for future use
+    struct {
+        uint32_t data[16];  // @user:       User data (optional)
+    } user;
+    uint32_t checksum;      // @checksum:   Partition entry checksum (includes all above)
+} __attribute__ ((packed));
+```
+
+Each partiton can contain image of following types:
+- FFS_TYPE_DATA
+- FFS_TYPE_LOGICAL
+- FFS_TYPE_PARTITION
+
+# Partition layout
+Data presented below was extracted using `./fcp -L file:flash.pnor -o 0` and
+enchanced with additional data extracted directly from `flash.pnor`.
+More on how to use this tool can be found in [porting.md#ffs-tools](porting.md#ffs-tools).
+
+Partition Table
+Version: 1
+Size: 2 * 0x1000 (block size)
+Block Count: 0x4000
+Entry Size: 0x80
+Number of entries: 0x21
+
+| Name                        | ID | Offset:End         | Size    | Type | Flags | Actual size  | coreboot usage              |
+| --------------------------- | -- | ------------------ | -------:| ---- | ----- | ------------:| --------------------------- |
+| [part](#part)               | 01 | 00000000:00001FFF  | 2000    | p    | p     | same as size | partition table             |
```

Comment:
I assume the size is in hex and then bytes? Let's provide such information at least. Maybe human readable unit would be more useful?

Reply:
Information about hex and unit is now present.